framework/math/ParallelVector/ParallelSTLVector.cc

```diff
@@ -124,104 +199,61 @@ void ParallelSTLVector::operator+=(const ParallelVector& y)
   ChiLogicalErrorIf(y.GlobalSize() != global_size_,
                     "y.GlobalSize() != global_size_");
 
-  for (size_t i = 0; i < local_size_; ++i)
-    values_[i] += y[scint64_t(i)];
-}
-
-void ParallelSTLVector::operator+=(const ParallelSTLVector& y)
-{
-  ChiLogicalErrorIf(y.LocalSize() != local_size_,
-                    "y.LocalSize() != local_size_");
-  ChiLogicalErrorIf(y.GlobalSize() != global_size_,
-                    "y.GlobalSize() != global_size_");
+  // This prevents us from unnecessarily going through the operator[]
+  // of y, which again applies bounds checking
+  const double* y_data = y.Data();
 
-  for (size_t i = 0; i < local_size_; ++i)
-    values_[i] += y.values_[i];
+  for (int64_t i = 0; i < local_size_; ++i)
+    values_[i] += y_data[i];
 }
 
-void ParallelSTLVector::PlusAY(double a, const ParallelVector& y)
+void ParallelSTLVector::PlusAY(const ParallelVector& y, double a)
 {
   ChiLogicalErrorIf(y.LocalSize() != local_size_,
                     "y.LocalSize() != local_size_");
   ChiLogicalErrorIf(y.GlobalSize() != global_size_,
                     "y.GlobalSize() != global_size_");
 
+  // This prevents us from unnecessarily going through the operator[]
+  // of y, which again applies bounds checking
+  const double* y_data = y.Data();
+
   if (a == 1.0)
-    for (size_t i = 0; i < local_size_; ++i)
-      values_[i] += y[scint64_t(i)];
+    for (int64_t i = 0; i < local_size_; ++i)
+      values_[i] += y_data[i];
   else if (a == -1.0)
-    for (size_t i = 0; i < local_size_; ++i)
-      values_[i] -= y[scint64_t(i)];
+    for (int64_t i = 0; i < local_size_; ++i)
+      values_[i] -= y_data[i];
   else
-    for (size_t i = 0; i < local_size_; ++i)
-      values_[i] += a * y[scint64_t(i)];
+    for (int64_t i = 0; i < local_size_; ++i)
+      values_[i] += a * y_data[i];
```

Comment:
See comment above about `std::transform`.

Reply:
I am going to defer this till I can create benchmarks. My experience has always been that the transform via a lambda is pretty cache-unfriendly.